### ISSUE TITLE
Logical bug with totalRecords and filteredRecords

### ISSUE THREAD

**Issue by intech** [yajra/laravel-datatables#333]
```php
$this->totalRecords = $this->count();
```
https://github.com/yajra/laravel-datatables/blob/master/src/Engines/BaseEngine.php#L587

```php
$this->filteredRecords = $this->isFilterApplied ? $this->count() : $this->totalRecords;
```
https://github.com/yajra/laravel-datatables/blob/master/src/Engines/BaseEngine.php#L642

totalRecords and filteredRecords are always equal in all conditions!
They use one function $this->count() which returns the filtered results...

**Comment by yajra** [comment]
Not sure if there is a bug here since filtered and total values works fine for me. Can you please submit a PR to fix this?

**Comment by intech** [comment]
Greetings!
https://datatables.net/faqs/index#Server-side-processing
![](http://i.snag.gy/ykM8r.jpg)

https://datatables.net/manual/server-side#Returned-data 
![](http://i.snag.gy/wKcB1.jpg)

Open this is example: https://datatables.net/examples/server_side/row_details.html

Query:
```javascript
...
order[0][column]:1
order[0][dir]:asc
start:0
length:10
search[value]:te
search[regex]:false
...
```

Results:
```javascript
{
data: [{DT_RowId: "row_3", first_name: "Ashton", last_name: "Cox", position: "Junior Technical Author",…},…],
draw: 7,
recordsFiltered: 14,
recordsTotal: 57
}
```
> Showing 1 to 10 of **14** entries (filtered from **57** total entries)

This is **14** - count filtered results with searching (**te**) and **57** is total without filtering.

I propose to introduce a separate function totalCount() to be in any of the engines comply with the request without filtering conditions.

**Comment by yajra** [comment]
Based on this [demo](http://datatables.app/eloquent/basic-object), it shows the same expected result when search "ne".

`Showing 1 to 10 of 125 entries (filtered from 500 total entries)`

<img width="1021" alt="screen shot 2016-01-09 at 3 35 41 pm" src="https://cloud.githubusercontent.com/assets/2687997/12215063/07772540-b6e7-11e5-8372-5860bbfd2d5b.png">

**Comment by intech** [comment]
What engine do you use? Link to the local name :)

**Comment by yajra** [comment]
Oops sorry! ^_^.. Here's the live [demo](http://datatables.yajrabox.com/eloquent/basic-object) using eloquent.

**Comment by intech** [comment]
https://github.com/yajra/laravel-datatables/blob/6.0/src/Engines/QueryBuilderEngine.php#L75
This engine is written with the terms of its own logic calculation.
I'm in your engine under ElasticSearch made a similar but this is not correct.
It is better to divide it into two different functions.

The new engine is necessary inherit from BaseEngine where counting is wrong.

**Comment by yajra** [comment]
Can you please submit a PR to implement your proposal? Quite busy at work projects. Thanks a lot!

**Comment by intech** [comment]
Oh sure! I had already done in its working draft.

**Comment by intech** [comment]
All the same, I agree that this is not a bug.
This architectural solution for a particular problem. We need to make it more versatile for the other engines. Thank you for your activity, together, we understand the issues :)

**Comment by intech** [comment]
Why this happened:

In the example of the query builder, we call the count() several times. Before and after the application of filters, thereby counting from the beginning of the totalRecords and filteredRecords after filtered.

In my case, no query builder. To take the totalRecords I need to make a very different request, nothing like getting the data. I have come filteredRecords from the database in response to the data itself.

**Comment by yajra** [comment]
Make more sense now. Thanks for the PR. Merge and released on v6.1.3. :100: 

BTW, any chance you may want your engine to be included in the package? It would be a great help to the community to have an ElasticSearchEngine for DataTables. ^_^

**Comment by intech** [comment]
Thanks! Yes, I was planning to do it later when I have code refactoring. :)
